libavformat/rtcenc.c

```diff
@@ -0,0 +1,2226 @@
+ * timeout. In such case, publishers can't republish the stream util the session
+ * is timeout.
+ * This function is called to notify the server that the stream is ended, server
+ * should expire and close the session immediately, so that publishers can republish
+ * the stream quickly.
+ */
+static int whip_dispose(AVFormatContext *s)
+{
+    int ret;
+    char buf[MAX_URL_SIZE];
+    URLContext *whip_uc = NULL;
+    RTCContext *rtc = s->priv_data;
+
+    if (!rtc->whip_resource_url)
+        return 0;
+
+    ret = ffurl_alloc(&whip_uc, rtc->whip_resource_url, AVIO_FLAG_READ_WRITE, &s->interrupt_callback);
+    if (ret < 0) {
+        av_log(s, AV_LOG_ERROR, "Failed to alloc WHIP delete context: %s\n", s->url);
+        goto end;
+    }
+
+    av_opt_set(whip_uc->priv_data, "chunked_post", "0", 0);
+    av_opt_set(whip_uc->priv_data, "method", "DELETE", 0);
+    ret = ffurl_connect(whip_uc, NULL);
+    if (ret < 0) {
+        av_log(s, AV_LOG_ERROR, "Failed to DELETE url=%s\n", rtc->whip_resource_url);
+        goto end;
+    }
+
+    while (1) {
+        ret = ffurl_read(whip_uc, buf, sizeof(buf));
+        if (ret == AVERROR_EOF) {
+            ret = 0;
+            break;
+        }
+        if (ret < 0) {
+            av_log(s, AV_LOG_ERROR, "Failed to read response from DELETE url=%s\n", rtc->whip_resource_url);
+            goto end;
+        }
+    }
+
+    av_log(s, AV_LOG_INFO, "WHIP: Dispose resource %s ok\n", rtc->whip_resource_url);
+
+end:
+    ffurl_closep(&whip_uc);
+    return ret;
+}
+
+static av_cold int rtc_init(AVFormatContext *s)
+{
+    int ret;
+    RTCContext *rtc = s->priv_data;
+
+    if ((ret = whip_init(s)) < 0)
+        return ret;
+
+    if ((ret = parse_codec(s)) < 0)
+        return ret;
+
+    if ((ret = generate_sdp_offer(s)) < 0)
+        return ret;
+
+    if ((ret = exchange_sdp(s)) < 0)
+        return ret;
+
+    if ((ret = parse_answer(s)) < 0)
+        return ret;
+
+    if ((ret = ice_handshake(s)) < 0)
+        return ret;
+
+    /* Now UDP URL context is ready, setup the DTLS transport. */
+    rtc->dtls_ctx.udp_uc = rtc->udp_uc;
+
+    if ((ret = dtls_context_handshake(&rtc->dtls_ctx)) < 0)
+        return ret;
+
+    if ((ret = setup_srtp(s)) < 0)
+        return ret;
+
+    if ((ret = create_rtp_muxer(s)) < 0)
+        return ret;
+
+    return ret;
+}
+
+static int rtc_write_packet(AVFormatContext *s, AVPacket *pkt)
+{
+    int ret;
+    RTCContext *rtc = s->priv_data;
+    AVStream *st = s->streams[pkt->stream_index];
+    AVFormatContext *rtp_ctx = st->priv_data;
+
+    /* TODO: Send binding request every 1s as WebRTC heartbeat. */
+    /* TODO: Receive packets from the server such as ICE binding requests, DTLS messages,
+     * and RTCP like PLI requests, then respond to them.*/
+
+    /* For audio OPUS stream, correct the timestamp. */
+    if (st->codecpar->codec_type == AVMEDIA_TYPE_AUDIO) {
+        pkt->dts = pkt->pts = rtc->audio_jitter_base;
+        rtc->audio_jitter_base += 960;
```

Comment:
Kieran Kunhya:

Correct how?

Reply:
For opus 48khz, each frame is 20ms which is 48000*20/1000 = 960. It appears that there is a bug introduced by libopus regarding the timestamp. Instead of being exactly 960, there is a slight deviation, such as 956 or 970. This deviation can cause Chrome to play the audio stream with noise.

Although we are unsure of the root cause, we can simply correct the timestamp by using the timebase of Opus. We need to conduct further research and remove this line.

看起来似乎是libopus的bug，opus每帧应该是20ms也就是pts增量是960，而有时候它是956或970总是带一点误差，这点误差会造成Chrome播放时有噪音，噗噗噗这种杂音。我没有找到根本原因，只是暂时绕过去了这个问题，重新计算了时间戳。这里需要做更多的调研后正确处理时间戳。